Add 'policies' key to profile YAML files

jan-cerny opened the pull request:
Add 'policies' key to profile YAML files

jan-cerny committed:
Add 'policies' key to profile YAML files

jan-cerny committed:
Move ResolvableProfile closer to its parent class

This will allow us to reuse it in other modules.

jan-cerny committed:
Use the ResolvableProfile class in our tests

jan-cerny committed:
Pass controls manager to be used in resolve method

jan-cerny committed:
Parse 'policies' YAML key during profile parsing

jan-cerny committed:
Resolve controls to rule selections

jan-cerny committed:
Add support for variables in profiles

jan-cerny committed:
Extend profiles

The selections in child profiles will add or remove rules from
the parent profile and override the values of XCCDF values.

jan-cerny committed:
Resolve controls during profile compilation

This will allow us to use the controls in our profiles and will
resolve the controls to the rulesduring the build process.

matejak commented on ssg/build_yaml.py:
```diff
@@ -130,6 +130,7 @@ def from_yaml(cls, yaml_file, env_yaml=None):
         if selection_entries:
             profile._parse_selections(selection_entries)
         del yaml_contents["selections"]
+        policies = yaml_contents.pop("policies", None)
```
I think that we could add a method `resolve_control` that would assume that we *somehow* got a policy-control pair, and it would translate it to selections, values and refinements using the manager.
I would love to have implementation details about how to get policies and controls in an another PR.

matejak commented:
I suggest splitting this PR in two, and merging one of those into the other PR.
All commits except `Add 'policies' key to profile YAML files ` and ` Parse 'policies' YAML key during profile parsing ` are independent from the `.profile` file schema, and we may need some flexibility there.

jan-cerny committed:
Add option to select all controls

Using special keyword 'all' instead of list of controls in profile.yaml
all the controls from given policy will be selected. This is practical
for profiles that implement the whole controls catalog.

jan-cerny committed:
Extract method resolve_controls

jan-cerny commented:
@matejak changed! See #5 .

matejak committed:
Increase flexibility of profile with policies implementation.

- Restructure profile classes, so their behavior can be extended by inheritance.
- Made ResolvableProfile generic profile that knows the concept of controls, without implementation details.
- Introduced ProfileWithInlinePolicies, renamed existing implementation to ProfileWithSeparatePolicies.
- Extended the test suite, so it is able to test more implementations in a scalable way.

jan-cerny committed:
Add abcd-all.profile to tests data

matejak committed:
Improvements

- Added test for the all-inline case.
- Stramlined the inheritance relation of profiles.
- Removed _control_to_items as it was not helpful.

matejak committed:
Streamlined controls processing.

- Prevented conversion from control to ID and to control again in _controls_to_items
- Removed _policy_to_items
- Modified the flow, so now controls are obtained in a list, and then merged to the profile one-by-one.

jan-cerny committed:
Merge pull request #6 from matejak/controls_process

Increase flexibility of profile with policies implementation.

jan-cerny committed:
Merge pull request #5 from jan-cerny/controls_process

Expand controls to specific rules and variables during profile resolution
